### PR TITLE
Prediction Metrics: New module

### PR DESCRIPTION
**Pull request by orhankislal**
JIRA: MADLIB-907

A collection of summary statistics to gauge model accuracy
based on predicted values vs. ground-truth values.

### CODE REVIEW

File: doc/mainpage.dox.in [hunk 1]
```diff
@@ -117,10 +117,11 @@ complete matrix stored as a distributed table.
         @ingroup grp_datatrans
 
 @defgroup grp_mdl Model Evaluation
-@{Contains the cross-validation module, a collection of routines useful for
-<a href="http://en.wikipedia.org/wiki/Cross-validation_(statistics)">Cross-validation</a>. @}
+@{Contains functions for ensuring accuracy and validation of predictive methods. @}
```
Comment (iyerr3) [inline doc/mainpage.dox.in]: I would use "evaluating" instead of "ensuring"

**Comment by iyerr3** [comment]
General comments: 
- The distance functions (mean_*_error) all have the same structure except the distance metric. I suggest refactoring the table creation and the grouping statements into a single function that takes a parameter for the distance. 
- Avoid using `__` within python. It leads to [name mangling](https://docs.python.org/2/tutorial/classes.html#private-variables-and-class-local-references), that doesn't really help in this situation. 
- There seems to be considerable overlap between the grouping and non-grouping SQL in `r2`, `adjusted_r2` and `auc` functions. Maybe we can combine them? 
- I'll bump the [suggestion](https://github.com/apache/incubator-madlib/pull/41#issuecomment-216661096) by @decibel in the previous PR. It might be helpful to have 2 versions: 1 that creates the output table and the other returns the rows directly.

**Comment by iyerr3** [comment]
I have made some changes and added validation and online help functions (in my [private fork branch](https://github.com/iyerr3/incubator-madlib/commits/feature/pred_metrics_take2)). 

I have not implemented the SRF versions as discussed earlier. We can make those functions as a separate work. 

@orhankislal could you please pull in the 3 commits on that branch and update your branch to update this PR. Once we've discussed the changes here, I can merge the commits to master.

File: src/ports/postgres/modules/stats/pred_metrics.py_in [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+        SELECT {grp_str}
+               threshold,
+               sum(t) OVER ({partition_str} ORDER BY threshold DESC) *
+                    1.0/ NULLIF(sum(t) OVER ({partition_str}), 0) AS tpr,
+               sum(f) OVER ({partition_str} ORDER BY threshold DESC) *
+                    1.0/NULLIF(sum(f) OVER ({partition_str}),0) AS fpr
+        FROM (
+          SELECT {grp_str}
+                 {pred_col} AS threshold,
+                 sum({obs_col}) AS t,
```
Comment (iyerr3) [inline src/ports/postgres/modules/stats/pred_metrics.py_in]: Let's add a `::int` inside the sum. That way we can also support `obs_col` as a boolean (which would be nice considering this is binary classification).

File: src/ports/postgres/modules/stats/pred_metrics.py_in [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+                sum(t) OVER ({partition_str} ORDER BY threshold DESC) AS tp,
+                sum(f) OVER ({partition_str} ORDER BY threshold DESC) AS fp,
+                sum(t) OVER ({partition_str}) -
+                    sum(t) OVER ({partition_str} ORDER BY threshold DESC) AS fn,
+                sum(f) OVER ({partition_str}) -
+                    sum(f) OVER ({partition_str} ORDER BY threshold DESC) AS tn
+            FROM (
+                SELECT {grp_str}
+                       {pred_col} AS threshold,
+                       sum({obs_col}) AS t,
```
Comment (iyerr3) [inline src/ports/postgres/modules/stats/pred_metrics.py_in]: Let's cast `obs_col` to int (also line below).

**Comment by iyerr3** [comment]
Along with casting the columns to int in binary classification, we also need to change docs/online-help/tests to reflect that boolean columns allowed for observation columns.

File: src/ports/postgres/modules/stats/pred_metrics.py_in [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+                SELECT {grp_out_str}
+                       sum(({obs_col} - mean)^2) as sstot,
+                       sum(({pred_col} - {obs_col})^2) AS ssres
+                FROM(
+                    SELECT {grp_out_str}
+                           {pred_col}, {obs_col},
+                           avg({obs_col}) OVER ({partition_str}) as mean
+                    FROM {table_in}
+                ) x {grp_by_str}
+            ) y
```
Comment (mktal) [inline src/ports/postgres/modules/stats/pred_metrics.py_in]: Why not 
```sql
SELECT 
     {grp_out_str}
     1 - avg(({pred_col} - {obs_col})^2)/var_pop({obs_col}) AS r2_score
FROM {table_in} {grp_by_str}
```
It is simpler, faster (2-3 times in my quick experiments) and numerically stable (avoid large sum)

File: src/ports/postgres/modules/stats/pred_metrics.py_in [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
+        FROM (
+            SELECT
+                {grp_str}
+                threshold,
+                sum(t) OVER ({partition_str} ORDER BY threshold DESC) AS tp,
+                sum(f) OVER ({partition_str} ORDER BY threshold DESC) AS fp,
+                sum(t) OVER ({partition_str}) -
+                    sum(t) OVER ({partition_str} ORDER BY threshold DESC) AS fn,
+                sum(f) OVER ({partition_str}) -
+                    sum(f) OVER ({partition_str} ORDER BY threshold DESC) AS tn
```
Comment (mktal) [inline src/ports/postgres/modules/stats/pred_metrics.py_in]: I suggest defining the window in the query's WINDOW clause, e.g., `WINDOW w AS ({partition_str} ORDER BY threshold DESC)` and just reference it in the query, e.g., `sum(t) over w AS tp`.

File: src/ports/postgres/modules/stats/pred_metrics.sql_in [hunk 6]
```diff
@@ -0,0 +1,831 @@
...
+- \f$\textit{f1}=2*\textit{tp}/(2*\textit{tp}+\textit{fp}+\textit{fn})\f$.
+
+
+<b>Area under ROC Curve</b>
+
+This function returns the area under the Receiver Operating Characteristic curve
+for binary classification (the AUC). The ROC curve is the curve relating the
+classifier's TPR and FPR metrics. (See <a href="#bc">Binary Classification</a>
+for a definition of these metrics). Please refer to the reference <a
+href="#aoc">[2]</a> for further details. Note that the binary classification
```
Comment (mktal) [inline src/ports/postgres/modules/stats/pred_metrics.sql_in]: Please check if the reference number is correct. [2] points to confusion matrix.

File: src/ports/postgres/modules/stats/pred_metrics.sql_in [hunk 7]
```diff
@@ -0,0 +1,831 @@
...
+it will be '0'
+
+
+<b>Confusion Matrix</b>
+
+This function returns the confusion matrix of a multi-class classification. Each
+column of the matrix represents the instances in a predicted class while each
+row represents the instances in an actual class. This allows more detailed
+analysis than mere proportion of correct guesses (accuracy). Please refer to the
+reference <a href="#cm">[3]</a> for further details. As noted earlier, grouping
```
Comment (mktal) [inline src/ports/postgres/modules/stats/pred_metrics.sql_in]: [3] is aoc

**Comment by orhankislal** [comment]
Thank you very much for your comments @iyerr3 and @mktal. I have pushed 2 new commits to incorporate your suggestions. Please let me know if you have any other suggestions.

**Pull request closed without merging**